src/main/java/graphql/schema/validation/ObjectsImplementInterfaces.java

```diff
@@ -114,4 +106,37 @@ private SchemaValidationError error(String msg) {
         return new SchemaValidationError(ObjectDoesNotImplementItsInterfaces, msg);
     }
 
+    boolean isCompatible(GraphQLOutputType a, GraphQLOutputType b) {
+        if (isSameType(a, b)) {
+            return true;
+        } else if (a instanceof GraphQLUnionType) {
+            return objectIsMemberOfUnion((GraphQLUnionType) a, b);
+        } else if (a instanceof GraphQLInterfaceType && b instanceof GraphQLObjectType) {
+            return objectImplementsInterface((GraphQLInterfaceType) a, (GraphQLObjectType) b);
+        } else if (a instanceof GraphQLList && b instanceof GraphQLList) {
+            GraphQLOutputType wrappedA = (GraphQLOutputType) ((GraphQLList) a).getWrappedType();
+            GraphQLOutputType wrappedB = (GraphQLOutputType) ((GraphQLList) b).getWrappedType();
+            return isCompatible(wrappedA, wrappedB);
+        } else if (b instanceof GraphQLNonNull) {
+            GraphQLOutputType wrappedB = (GraphQLOutputType) ((GraphQLNonNull) b).getWrappedType();
+            return isCompatible(a, wrappedB);
+        } else {
+            return false;
+        }
+    }
+
+    boolean isSameType(GraphQLOutputType a, GraphQLOutputType b) {
+        String aDefString = getUnwrappedTypeName(a);
+        String bDefString = getUnwrappedTypeName(b);
+        return aDefString.equals(bDefString);
+    }
+
+    boolean objectImplementsInterface(GraphQLInterfaceType interfaceType, GraphQLObjectType objectType) {
+        return objectType.getInterfaces().contains(interfaceType);
+    }
+
+    boolean objectIsMemberOfUnion(GraphQLUnionType unionType, GraphQLOutputType objectType) {
+        return  unionType.getTypes().contains(objectType);
+    }
```

Comment:
the contains here is going to be object equality since they dont have hashCode / equals methods